### PR TITLE
check for kodi.inputstream version

### PR DESCRIPTION
**Pull request by ghost**
see title

### CODE REVIEW

**Comment by MilhouseVH** [comment]
This does the job here, many thanks!

Logging when there is an API mismatch might be a good idea, but not essential. We will include this as a kodi 17 beta1 backport in LibreELEC.

File: xbmc/addons/InputStream.cpp [hunk 1]
```diff
@@ -36,8 +36,14 @@ std::unique_ptr<CInputStream> CInputStream::FromExtension(AddonProps props, cons
   std::string extensions = CAddonMgr::GetInstance().GetExtValue(ext->configuration, "@extension");
   std::string protocols = CAddonMgr::GetInstance().GetExtValue(ext->configuration, "@protocols");
   std::string name(ext->plugin->identifier);
+
+  bool matchAPIVersion(true);
+  for (unsigned int i(0); i < ext->plugin->num_imports; ++i)
+    if (strcmp(ext->plugin->imports[i].plugin_id,"kodi.inputstream")==0 && strcmp(ext->plugin->imports[i].version, INPUTSTREAM_API_VERSION))
+      matchAPIVersion = false;
```
Comment (da-anda) [inline xbmc/addons/InputStream.cpp]: shouldn't we break the loop once we found the desired plugin_id to save a couple CPU cycles?

File: xbmc/addons/InputStream.cpp [hunk 2]
```diff
@@ -36,8 +36,14 @@ std::unique_ptr<CInputStream> CInputStream::FromExtension(AddonProps props, cons
   std::string extensions = CAddonMgr::GetInstance().GetExtValue(ext->configuration, "@extension");
   std::string protocols = CAddonMgr::GetInstance().GetExtValue(ext->configuration, "@protocols");
   std::string name(ext->plugin->identifier);
+
+  bool matchAPIVersion(true);
+  for (unsigned int i(0); i < ext->plugin->num_imports; ++i)
+    if (strcmp(ext->plugin->imports[i].plugin_id,"kodi.inputstream")==0 && strcmp(ext->plugin->imports[i].version, INPUTSTREAM_API_VERSION))
```
Comment (fritsch) [inline xbmc/addons/InputStream.cpp]: Please add != 0 to the second statement, as people not knowing what strcmp returns lead to confusions.

**Comment by FernetMenta** [comment]
-1
no cpluff hacks for developer versions. absolutely no need for this.

**Comment by MilhouseVH** [comment]
@FernetMenta I know where you're coming from and I'm sympathetic to that (and I don't even care if this isn't merged as we'll still use it) but please understand you now have thousands of installations that can't be upgraded easily because of this issue.

They may be testers and this may be alpha, but they all give their spare time just like everyone else working on Kodi so I don't think we should dump on them if we can avoid it, otherwise we'll have a lot less testers in future.

Please at least review this as if it were to be merged, even if ultimately it isn't merged as then it's down to distributions to carry this patch temporarily until test users have migrated to a non-crashing combination of Kodi 17 and inputstream.

**Comment by FernetMenta** [comment]
I reviewed it and I already told you that I did think about a solution. But this change is a hack and should not be merged to master. LE can include it and drop it later.
It is appreciated that all those users contribute by testing alpha versions but we can't introduce a fully functional versioning system for developer branches. Every single hack like this will hit you at a later time with even more work to correct it.

**Comment by mapfau** [comment]
I think it could be a good solution to check dependencies always on load time of whatever addons. Not only inputstream. Then the chance that kodi doesnt segfaults on version mismatch is much greater.
Unfortunately i haden't the time to look deeper up the stack trace.

From my personal experience on kodinerds: tons of hours were spend because of exactly this issue.
There are plenty of posts because of kodi crash in the time the api has changed. A nice log entry wich explains this would be better than a stack trace (log entry is missing here, sry).

If you @FernetMenta agree to the idea of a solution for this wich is not hacky and more generic, I culd look deeper. If you don't want such a checkforever reason, I can save my time.... Pls. let me know.

**Comment by MartijnKaijser** [comment]
Doesn't https://github.com/xbmc/xbmc/pull/10335 partially solve it? At least when it's properly set in addon.xml

**Comment by MilhouseVH** [comment]
@martihn my latest master build that includes #10335 continues to crash if the user has an older inputstream addon installed, so no (at least, not in #10335's current state).

**Comment by tamland** [comment]
Consider #10335 a wip, but the proper solution for inputstream is to build on that and make something general that works for the entire addon system; not just adding even more binary addon/inputstream specific handling.

**Comment by mapfau** [comment]
I'll look into the PR today

**Comment by AlwinEsch** [comment]
Related to version check, I has also another way added on my rework (related to a API level system, not to a Version string), see https://github.com/AlwinEsch/kodi/commit/104a4ecca49e325a1b0a71c8f3636c1abeaa38fb

Not sure if my way becomes better, but another way.

**Comment by mapfau** [comment]
@AlwinEsch this means that the api level is already defined in addons. Unfortunately the child has already fallen into the fountain. We try to fix already installed addons.

My knowledge about kodi versioning is not quite good, but I think there is a simple check of all the requirements missing. This should be checked before something is loaded.

My approach would be collecting all "importable" kernel modules in a map, and then let AddonLoader or DLLLoader or CPLuff or whatever makes sense let check all import statements

This one for example.
```
  <requires>
    <import addon="xbmc.core" version="0.1.0"/>
  </requires>
```
would be one wich has to be in the map. Unfortunately I'm not able to find some #define of this xbmc.core version in the kodi sources. Anybody an idea?

**Comment by MartijnKaijser** [comment]
@mapfau is this still needed or obsolete with Addon manager rework of week ago?

**Comment by mapfau** [comment]
closed -> it was just an idea

**Pull request closed without merging**